### PR TITLE
[rust] Tracking Selenium Manager usage through Plausible (#11211)

### LINKED ISSUE #11211: [🚀 Feature]: Implement download tracking in selenium manager
### Feature and motivation

The idea is to get basic usage information about who is using Selenium Manager. 

We could create a basic page on `https://www.selenium.dev/selenium/manager`. The Selenium Manager would route all requests for downloads through that page.

Something like:
```
https://selenium.dev/selenium/download/manager?language=ruby&os=linux&driver=geckodriver&driver_version=v0.32.0
```

Which would redirect and return the correct file, but would allow us to see what language/os/browser/browser versions are being used.

### PR DESCRIPTION
**Pull request by bonigarcia**
### Description
This PR implements a mechanism to track the usage of Selenium Manager through [Plausible](https://plausible.io/). 

To that, the Rust code sends a `pageview` request to plausible using the events API to track visitors. As requested in #11211, the data we want to collect is the following:

- Selenium version
- Language binding 
- Browser
- Browser version
- Operating system
- Architecture

For "language binding", a new flag called `--language-binding` has been introduced in this PR. Bindings should use this property to report it (e.g., `--language-binding java`, `--language-binding ruby`, etc.).

All of this information is gathered through custom properties (called `props` in the [Plausible Event API](https://plausible.io/docs/events-api)). The (fake) page to track these visitors will be the following:

https://selenium.dev/sm-stats

An example of the equivalent POST request to track a page view is the following:

```
curl -i -X POST https://plausible.io/api/event \
  -H 'User-Agent: Selenium Manager 0.4.16' \
  -H 'Content-Type: application/json' \
  --data '{"name":"pageview","url":"https://selenium.dev/sm-stats","domain":"selenium.dev","props":{"browser":"chrome","browser_version":"119.0.6045.105","os":"linux","arch":"x86_64","lang":"java","selenium_version":"4.16"}}'
```

The data gathered by Plausible will be available on:

https://plausible.io/selenium.dev

To configure the custom properties (e.g., for filtering), these properties should be enabled:

https://plausible.io/selenium.dev/settings/properties

I tested this feature using a free Plausible account and my personal domain  (bonigarcia.dev):

image

This PR is already pointing to selenium.dev. @diemol @titusfortner If you want to check how data is gathered, you can trigger the [CI Rust workflow](https://github.com/SeleniumHQ/selenium/actions/workflows/ci-rust.yml) using the branch `sm_plausible`.

NOTE: I was concerned about the impact of this feature on the overall performance of Selenium Manager. To get some numbers, I calculated the average time of running the Selenium Manager tests using the `trunk` branch (i.e., without this feature) compared to the branch `sm_plausible`. 

For that, I executed the following command in Linux:

```
time cargo test
```

I executed this command (with every asset in the cache) 10 times for each branch, and I calculated the average total time.

average_total_time_trunk = 11.1399 seconds
average_total_time_plausible = 12.4713 seconds

Therefore, the overhead introduced to the requests made to Plausible averages 0.7341 seconds. Since 23 page views were requested during the tests, each `pageview` request added an overhead of around 32 ms, which seems reasonable.

### Motivation and Context
This PR resolves #11211.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [X] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: rust/src/lib.rs
Comment (diemol) [inline rust/src/lib.rs]: Is this call blocking, or is it sent in a different thread? We do not want to block the main thread to send this. 

Also, we need an environment variable to allow users to opt-out.

Comment (bonigarcia) [reply rust/src/lib.rs]: > Is this call blocking, or is it sent in a different thread? We do not want to block the main thread to send this.

It's a blocking call. Everything happens in the same thread in Selenium Manager. But I can explore how to spawn a new thread for this call.

> Also, we need an environment variable to allow users to opt-out.

I was not aware it was a requirement. But I can implement it.

Comment (diemol) [reply rust/src/lib.rs]: >I was not aware it was a requirement. But I can implement it.

Yeah. We did not finish the conversation on #11211 before this PR was created. When implementing tracking, we need to give users the possibility to opt-out. 

>It's a blocking call. Everything happens in the same thread in Selenium Manager. But I can explore how to spawn a new thread for this call.

We need to make it non-blocking because it should not affect the user if this call fails.

Comment (bonigarcia) [reply rust/src/lib.rs]: I have just added two commits to this PR to implement the requested features. Now, the POST request to plausible.io is done asynchronously. Also, the argument `--avoid-stats` is accepted by Selenium Manager for not sending the statistics to plausible.io. As usual, this argument can be configured using the configuration file or an env (`SE_AVOID_STATS=true`).

**Comment by titusfortner** [comment]
Excellent. Have we tested the output? What does it look like on Plausible?

**Comment by titusfortner** [comment]
@bonigarcia is this ready?
@diemol do we see what the output looks like in Plausible?

**Comment by bonigarcia** [comment]
> is this ready?

Yes, it is.

**Comment by diemol** [comment]
@bonigarcia 

>NOTE: I was concerned about the impact of this feature on the overall performance of Selenium Manager. To get some numbers, I calculated the average time of running the Selenium Manager tests using the trunk branch (i.e., without this feature) compared to the branch sm_plausible.

Is all still the same after you have placed the call in a non blocking thread?

**Comment by bonigarcia** [comment]
> Is all still the same after you have placed the call in a non blocking thread?

The call to Plausible is now executed in a different thread, so the time overhead is null.

File: rust/src/main.rs
Comment (diemol) [inline rust/src/main.rs:226]: Could we move the stats creation and upload at the end? The "primary" tasks of SM should go first and these type of things at the end, ideally to prevent any issues. 

I mean, if there is an issue with the `stats()` code, would the execution continue?

Comment (bonigarcia) [reply rust/src/main.rs:226]: The call to Plausible (`.stats()`) is done in a different thread. This way, even if Plausible is offline or any other unexpected error happens, the regular behavior of Selenium Manager will not be affected since the `.setup()` logic is executed in the main thread. I have tested it.

**Comment by titusfortner** [comment]
@bonigarcia can you resolve conflicts here?

**Comment by bonigarcia** [comment]
> can you resolve conflicts here?

Done.

File: rust/src/main.rs
Comment (diemol) [reply rust/src/main.rs:226]: 👍

File: rust/src/stats.rs
Comment (diemol) [inline rust/src/stats.rs]:
```suggestion
const SELENIUM_DOMAIN: &str = "manager.selenium.dev";
```

**Review by diemol — COMMENTED** [review]
I changed the domain because I created one to only have the SM stats and avoid mixing them with the website.

image

**Comment by bonigarcia** [comment]
> I changed the domain because I created one to only have the SM stats and avoid mixing them with the website.

Great.

File: rust/src/stats.rs
Comment (diemol) [inline rust/src/stats.rs:49]: I know we did not mention this before, but... should we add the driver version? What do you think @bonigarcia?

Comment (bonigarcia) [reply rust/src/stats.rs:49]: I suppose it is not very relevant since most users would use the browser version (`browserVersion` set through options) but not the driver version.

BTW, the browser version sent to plausible is that value, i.e., what the user specifies through options (e.g., `"beta"`, `"120"`, `""`) and not the discovered browser version (e.g., `120.0.6099.71`, etc.). Is that what we want, right?

Comment (diemol) [reply rust/src/stats.rs:49]: I think it makes sense to avoid the driver version then.

Regarding the browser version... I think we should collect both. Does `requested_browser_version` sound good?

Comment (bonigarcia) [reply rust/src/stats.rs:49]: It can be done (gather both `requested_browser_version` and `browser_version`). But if we do that, we should send the stats after executing the driver/browser management. In other words, the call to `.stats()` is done after `.setup()`.

In that case, and since after the `.setup()` SM is almost finished (it only needs to log the output), we should wait for the `.stats()` to be finished (since now it is executed in a separate thread). Alternatively, we should execute both `.stats()` and `.setup()` in the same thread (as it was done in the original PR). In any case, the call to Plausible will include some ms overhead (as reported in the description of this PR).

Do we want to do that?

Comment (diemol) [reply rust/src/stats.rs:49]: That is a very good point. 

This also means that, with the current implementation, most of the time the `browser_version` value is empty because users normally don't specify a browser version.

To avoid loosing data, what if we execute `.stats()` in the same thread but shorten the read timeout to 1 second? In theory, we should not need to wait for a response. 

One last thing I noticed is that executing `.stats()` is not guarded by `selenium_manager.is_offline()`. I mean, if the user wants an offline execution, we should not send stats either.

Comment (titusfortner) [reply rust/src/stats.rs:49]: Can't we create the second process at the end of the setup() and not worry about the timeout?

Comment (diemol) [reply rust/src/stats.rs:49]: The problem is that the binary will exit and we might not send the POST completely.

Comment (titusfortner) [reply rust/src/stats.rs:49]: ah, so we're just taking advantage of the fact that setup() takes longer than stats() with the current implementation?

Comment (diemol) [reply rust/src/stats.rs:49]: I would say so.

Comment (titusfortner) [reply rust/src/stats.rs:49]: I actually think it is ok not to have discovered version. All that tells us is what the user happened to have installed on their machine? I don't think it is worth the effort to capture it...

Comment (diemol) [reply rust/src/stats.rs:49]: Why not? I think it is not a big effort, we are just trying to be careful to avoid blocking the user. I think the data is worth to have.

Comment (titusfortner) [reply rust/src/stats.rs:49]: My point is that if it adds a second to each test it's not worth the tradeoff. I'm more interested in what people request than what they have installed. Most likely what we'll see installed is the latest version of chrome because it auto updates.

Comment (titusfortner) [reply rust/src/stats.rs:49]: Actually, this is a good point. Enabling Selenium Manager on Grid is going to look different than using it in client. Can we differentiate when grid is requesting and when the Java client is rquesting?

Comment (diemol) [reply rust/src/stats.rs:49]: It does not add a second to each test, it should be a few ms as Boni pointed out. Worst case, it would add 1 second. We could even shorten that to 500ms.

Comment (bonigarcia) [reply rust/src/stats.rs:49]: > One last thing I noticed is that executing `.stats()` is not guarded by `selenium_manager.is_offline()`. I mean, if the user wants an offline execution, we should not send stats either.

Good catch. I would add that.

> ah, so we're just taking advantage of the fact that setup() takes longer than stats() with the current implementation?

Exactly.

> Actually, this is a good point. Enabling Selenium Manager on Grid is going to look different than using it in client. Can we differentiate when grid is requesting and when the Java client is rquesting?

I would say that the Grid implementation could use the flag `--language-binding` to report it. Perhaps: `--language-binding grid`.

Comment (titusfortner) [reply rust/src/stats.rs:49]: Can we just fire the event and not wait for the response since we aren't going to do anything with it, anyway?

Comment (bonigarcia) [reply rust/src/stats.rs:49]: I like the current implementation because it does not impact the overall performance. Moving the `.stats()` call to the end of the SM logic will affect some ms each time SM is executed. It is worth mentioning that in my tests (i.e., using the GHA infrastructure), this overhead is an average of 32 ms. But for any user, this figure will depend on the underlying network latency.

I tried to implement a *fire and forget* approach for the POST requests to Plausible in the Rust logic. But if the client doesn't wait for the response, Plausible does not count it. So, for gathering the discovered browser version (I can also include the found driver version for the same prize), I'm afraid we need to wait for the response, adding some MS overhead per test.

Comment (bonigarcia) [reply rust/src/stats.rs:49]: I have just added a commit to check the offline mode.

Please let me know your final decision about the current discussion. There are two alternatives:

a) Move the call to `stats()` at the end of SM logic and add the discovered browser version (and found driver version if you want) to the data sent to Plausible. This call should be blocking, and so, it will add a little overhead.
b) Leave it in the current way, in which the browser version is the one reported by the user and `stats()` is done at the beginning (concurrently).

Comment (diemol) [reply rust/src/stats.rs:49]: I think we could go with option a) and iterate. Maybe we do not need a perfect solution now, just something that helps us understand usage.

Comment (titusfortner) [reply rust/src/stats.rs:49]: My vote is to keep existing implementation and just get the browser-version passed by the user because that's the least impact.

Comment (titusfortner) [reply rust/src/stats.rs:49]: "usage" comes from what the user passes in. Knowing what a user has on their system seems extra. I think we should start with the least-impact and add to it if we think we need more details.

Comment (diemol) [reply rust/src/stats.rs:49]: 🤦 

I meant option `b)`, so as it currently is.

Comment (bonigarcia) [reply rust/src/stats.rs:49]: My vote, if I have one, is b. For me, it is better if this feature is not impacting the overall performance, even if it is a few ms.

Comment (diemol) [reply rust/src/stats.rs:49]: @bonigarcia,  of course you have a say!

Comment (titusfortner) [reply rust/src/stats.rs:49]: And we all agree, lol. We can definitely revisit after we start seeing data.

Comment (diemol) [inline rust/src/stats.rs:49]: Can we lower case all these properties, please? I just tried it and we get different values if they are sent with different casing.

**Review by diemol — CHANGES_REQUESTED** [review]

File: rust/src/stats.rs
Comment (diemol) [reply rust/src/stats.rs:49]: e.g.

image

Comment (bonigarcia) [reply rust/src/stats.rs:49]: > Can we lower case all these properties, please?

I have committed a patch to send them in lowercase.

Comment (diemol) [reply rust/src/stats.rs:49]: Great! Let's merge this and I will implement the extra flag in the different languages.

**Review by diemol — APPROVED** [review]
Thank you, @bonigarcia!

**Comment by zfouts** [comment]
What's the config option to opt out?  I don't like the idea of this doing an API call every time I run selenium. I've blocked it at my edge, but would prefer if we can just straight up disable the API calls.

**Comment by titusfortner** [comment]
Oh, I never posted the blog announcing the release
https://github.com/SeleniumHQ/seleniumhq.github.io/pull/1564/files

Environment variable SE_AVOID_STATS or with the config file

**Comment by titusfortner** [comment]
And for what it's worth it only reports usage once a day by default based on the ttl setting.

**Comment by bonigarcia** [comment]
> And for what it's worth it only reports usage once a day by default based on the ttl setting.

That is not correct. TTL is used only to give validity to the metadata. The metadata contains the knowledge of the proper driver version for a given browser version based on the response of online endpoints (e.g., CfT). The idea of using the TTL period for sending the stats to Plausible was proposed somewhere, but that decision was not made.

**Comment by rjhancock** [comment]
1) I wish this was communicated more publicly before it was put in place. I do understand the desire and need to know the information, but make it opt-in vs opt-out.

2) The method described for opting out appears to be a no-op unless it is applied in a library that is not included within the entire SeleniumHQ organization [as it only appears thrice](https://github.com/search?q=org%3ASeleniumHQ+SE_AVOID_STATS&type=code) and not a single place to disable sending actually occurs within the code.

3) It sends those reports multiple times. Just within my test suite it has attempted to send the info at least a dozen times across 1800 tests. Since it is also used in production, it would be reporting thousands of times A DAY just from our use case.

**Comment by titusfortner** [comment]
We're going to reduce the frequency of reports in the next version, but I'm not sure I understand your underlying concern.

**Comment by rjhancock** [comment]
The underlying concern is some of us use Selenium in more sensitive networks and now we have more work to protect the privacy of what we do to disable this "feature" that was opt-out vs opt-in.

I understand the desire to collect such metrics but a more diplomatic way of doing it, in the face of ever growing invasive privacy and tracking issues, would have been "we've added telemetry collection on an opt-in basis. It would help us to know where to focus our efforts if you'd turn it on" vs "<no mention except a single line in a readme file>"

Currently downgraded the version used until I can put in a place to permanently block all telemetry data as the above mentioned environment variable doesn't seem to be checked within the code anywhere and doesn't appear to be a known one for other systems either.

**Comment by diemol** [comment]
Regarding privacy, it is worth mentioning that we use https://plausible.io/, which is privacy first and GDPR compliant.

**Comment by titusfortner** [comment]
Thanks for clarifying. I was a little confused because the number of times the same information is sent seemed irrelevant to the point you were making.

Three potential concerns as I see it.

1. Performance — shouldn't be affected the http call made by a separate thread
2. Privacy — there is nothing personally identifiable, browser name, version, os, architecture, calling language. Plausible itself gathers IP addresses for locations, but it is only aggregated: https://plausible.io/privacy-focused-web-analytics
3. Security — If there is a proxy or other network restrictions, the call won't go anywhere. Also we're still working on getting these signed

We talked about a several different ways to implement this, some more complex and some more simple, and this was the best compromise we came up with. The binary itself has not been opt-in for a while now. The new behavior (automatic downloads) is currently opt-in, but this will change soon as well.

**Comment by rjhancock** [comment]
It seems my concerns weren't clear enough. Regardless of Plausible's stance on privacy, that does not mean all of us wish to share such data for a variety of reasons and this change caused several alarms to flag on our end during testing due to unexpected requests being made that we did not authorize.

I don't deny that the data you request is valuable for the project as a whole, but such analytics should be Opt In vs Opt Out. Ask the community as a whole to opt-in to new analytics and tell them why, don't just turn it on with barely a mention in the readme.

This issue caused additional billable time to my client to ensure the data is not sent so this decision is having monetary damages to individuals and companies.

**Comment by walksonair** [comment]
Thank you for providing visibility into this metric gathering. Due to some sensitive work/contracts, I can't allow any data to escape. Could you let me know if there's a way to turn this off for such projects that require it? At the moment, we get `Error sending stats to Plausible` on some networks. Again, it would help if we could opt out of this metric collection for some projects. Thank you again!

**Comment by titusfortner** [comment]
I understand your concerns, and at the very least it could have been communicated better. Opt-in is typically for personally identifiable info or tracking data; we view this as essentially the same GDPR compliant data we collect by default from all of our website visitors. Everything collected is publicly available. If we had more resources and more assistance we might have found other ways of implementing or communicating things, but we're doing the best we can do with the resources we have.

**Comment by titusfortner** [comment]
Yes, you can disable it by setting environment variable SE_AVOID_STATS to "true" or with the [config file](https://www.selenium.dev/documentation/selenium_manager/#configuration)

**Comment by rjhancock** [comment]
We've found that using the environment variable does somehow work despite not being checked in code and are using that IN ADDITION to other methods.

I'm not dismissing the work you're doing, we do appreciate it, but even with it being GDPR compliant collection it is still something that should have been opt-in.

**Comment by titusfortner** [comment]
> despite not being checked in code

It is checked by the binary (generated from the Rust code). 
We decided that it was only worth collecting the data if it was opt-out. 
That there would be concerns and frustrations was foreseeable and we could have communicated it better. 

@diemol / @bonigarcia do either of you have time to write a blog post and update our docs? I can add it to my list if not.

**Comment by zfouts** [comment]
@titusfortner FYI on the release blog post, it was categorized as 2023 by mistake, may wish to move that to the 2024 folder so it's more obvious and visible. 

I agree with others that it should have been communicated more widely in advance, and also that the data collected is valuable. However, an opt-in approach is always best when collecting data. But primarily, communication on the intent should have been made before the release.

**Comment by rjhancock** [comment]
> > despite not being checked in code
> 
> It is checked by the binary (generated from the Rust code).

Using GitHub search only shows it in [three places](https://github.com/search?q=org%3ASeleniumHQ+SE_AVOID_STATS&type=code) and looking through what some would consider to be obvious places does not show that environment variable being checked anywhere.

Even extending to all of GitHub still does not show where it is being checked.

Again, I agree the data is valuable in general and for the future of the project, and I do appreciate the hard work y'all put into this.

In a world where people don't like being tracked and reported on, especially developers, opt-in is ALWAYS the best approach EVEN when the data is compliant with all laws. The data being shared via a browser making a request is expected, that same data being harvested and sent to a third-party we didn't agree to, even if benign, is not.

**Comment by titusfortner** [comment]
I'm not sure why it matters where in the code it is set. 
it parses environment variables with `SE_*_*` and cli values with `--*-*` and config file values, etc
https://github.com/SeleniumHQ/selenium/blob/selenium-4.17.0/rust/src/lib.rs#L1478-L1480
Everything we see is publicly available on plausible's site - https://plausible.io/manager.selenium.dev

I'm not disagreeing with you, your viewpoint is valid, but all software is compromises and tradeoffs based on resources and requirements and the best I can say is that this implementation was at the confluence of our consensus evaluation.

**Comment by rjhancock** [comment]
The location matters to those of us who may now have to do additional auditing to ensure the dependencies we include in our projects don't introduce undesired effects such as this feature.

Might I suggest changing the default to be `true` until the next major version and re-communicate to the community why the change was made and why you want the data?

You still get your data, eventually, and can do a re-do on the communication front.

Every decision is a compromise and a set of trade offs regardless of it is software or IRL. Having this be opt-out for tracking was a bad decision.

**Comment by titusfortner** [comment]
Alternately, if you specify the location of the driver you want in the service class, the binary won't be used at all.

**Comment by rjhancock** [comment]
For those using the Ruby gem, configure a custom browser and pass an options package to it with this argument.

`options.add_argument('--avoid-stats=true')`

Still suggest putting in extra measure on your network to block the tracking as a backup however.

**Comment by titusfortner** [comment]
> pass an options package to it with this argument

That... shouldn't work. Selenium Manager needs the args and those options arguments get passed to the driver when starting the browser.

**Comment by rjhancock** [comment]
Oddly it does. Would love a more elegant solution to ensure this feature is disabled on deployments. Currently have a custom hosts file, environment variables set, and this to disable sending of data using the selenium-webdrivers gem.

**Comment by titusfortner** [comment]
Hmm, really think is has to be something else and/or this behavior will get changed unexpectedly for you. Can you turn on logging (`Selenium::WebDriver.logger.level = :debug`) and link to a gist with the results?

**Comment by rjhancock** [comment]
I saw it report it as unrecognized but it does not attempt to send the analytics.

https://gist.github.com/rjhancock/718994edf7756392336f07ae1a05d950

This is only 8 of 1800+ tests.

**Comment by andrew-azarov** [comment]
> Regarding privacy, it is worth mentioning that we use https://plausible.io/, which is privacy first and GDPR compliant.

It's not compliant, because when you install the package consent is not triggered. Thus when someone installs without reading README but by using standard Python tools they don't see a warning nor consent on stealing sensitive data (and just having an app receive IP address of client without explicit consent is enough for breaking PI laws).

As long as you can install selenium without triggering EXPLICIT consent - it is not legal.

Moreover - there are use cases when subdomain (non public) can be a privately identifiable (and remember in GDPR Article 4 is identifiable) information.
For example you setup a catchall subdomain ala jack-coch-zx0vu80.clients.website.tld for ease of access to a private panel for customer and limit that to 1 ip address or subnet - this is by all definitions private space and private identifiable information as nobody but the script runner can see it and access it. And now this goes to plausible without ANY consent.

Besides that neither manual nor README includes a statement about telemetry.
EDIT: not even website has consent or Privacy policy. This must be reported asap to national PII protection services.
Any company which uses selenium - could have inadvertently compromised their own privacy policies and must report breach (https://commission.europa.eu/law/law-topic/data-protection/reform/rules-business-and-organisations/obligations/what-data-breach-and-what-do-we-have-do-case-data-breach_en)

I see lawsuits written all over this

**Comment by diemol** [comment]
@andrew-azarov I am not a lawyer, but SFC approved this tool mainly because of its focus on data privacy.

> Built with privacy of your visitors in mind. No need to annoy them with a cookie/GDPR consent banner. 

Check their content about it. https://plausible.io/privacy-focused-web-analytics

**Comment by andrew-azarov** [comment]
> @andrew-azarov I am not a lawyer, but SFC approved this tool mainly because of its focus on data privacy.
> 
> > Built with privacy of your visitors in mind. No need to annoy them with a cookie/GDPR consent banner.
> 
> Check their content about it. https://plausible.io/privacy-focused-web-analytics

The issue is that USING the PI collecting tool makes you liable for the data you collect. So there should be liability disclosure ie privacy policy, BEFORE you collect the data, doesn't matter how or for what, it can be a local app collecting mouse movement on your PC without sending it to a third party - it still must disclose that this data is collected and saved so the user is aware that this happens. Otherwise it is by all definition a malware.
Just a note about data collection deep in the help docs is not enough, especially when you can completely skip the docs when installing via packages.

This is the first breach of GDPR.
Second one is that they SEND the data to plausible, inadvertently exposing IP addresses, library and OS data (which can include hostname in some cases, even hidden hostnames like Kernel compilation hostname of internal VM in case of BSD for example).
As there is no detailed description of what data is sent I can only speculate, but per GDPR if the bulk of data is enough for personal identification - it is not compliant, it must be anonymized beforehand.
Plausible in this case receives IP (because selenium connects to it) and browser, OS, whatever else there. It's enough to identify a person even without explicit personal information available like in case of hostname I wrote about previously.

So what should have been done.
Selenium MUST have a privacy policy with clearly described and written PI it uses. Also this document must disclose all of the third parties processing that data. The consent must be explicitly taken upon installation of software, if no consent and selenium wants to keep telemetry on by default - then selenium MUST NOT be able to install.

There is a reason Google switched to consent 2.0 recently. That is because GDPR was written by lawyers and not technical people, and lawyers don't understand that even ping command exposes your IP address.

**Comment by diemol** [comment]
We looked at https://plausible.io/data-policy, and that gave us an understanding of what is done with the data.

We have also placed this in the docs: https://www.selenium.dev/documentation/selenium_manager/#data-collection, but from what you are saying, is it not enough?

Again, I am not a lawyer, and we rely on the advice of others at SFC. However, would you be willing to help us improve based on your comments? We just want to make Selenium better by understanding global usage; we are not interested in individual data. I am just confused because the tone you are using is kind of threatening, so I wonder if you would like to help us improve.

**Comment by Schnurres** [comment]
So you implemented and launched a data collection feature without ever consulting a lawyer or data protection expert, but claim it is GDPR compliant without ever having verified this?

That seems very negligent to me. Simply saying 'I am not a lawyer' is not a solution either. When you implement such critical features, you have to ensure compliance before making any claims.

**Comment by diemol** [comment]
I did not say that; I said the tool was reviewed by [SFC](https://sfconservancy.org/), which owns Selenium and has a legal team. Could you tell me why you suggested that when I did not say that explicitly?

I am having trouble understanding the threatening tone. If there is something that needs improvement, we can work it out with your help. This is the type of behavior that burns down OSS maintainers; instead of collaborating, we get harsh comments.

**Comment by rjhancock** [comment]
As I stated previously, the issue is that data started being collected and sent without informed consent from us developers and the only notice that was given was a line item on the release notes.

It was not clearly stated what was going on or what was going to happen. Having it be an opt-in situation vs opt-out would have solved this problem AS WELL AS a plea to the community for why you want the data and requesting they turn it on.

I recently had this discussion with another OS project about adding Sentry monitoring. Everyone was on board for including the package as it will help, but we decided to very strictly have it off by default. We've already started notifying users MONTHS before we implement it, why we want to collect the data, and requesting they turn it on.

In a world where data is king, and is generally taken without consent, be the bigger firm and specifically REQUEST it BEFORE taking it. I've wasted several hours ensuring all of my systems have extra protections in place because I've yet to find a suitable replacement.

**Comment by titusfortner** [comment]
No PII is retained, and we make all information we collect public — https://plausible.io/manager.selenium.dev
If you can find lawyers that think they can win an argument that this constitutes a violation of personal data retention, they can talk to our lawyers.

I do understand your frustration; I also wish there were a better option, but there is no way to get meaningful information and have it be opt-in. So the options were either not to collect any information or to do so the way we have. I get that you would make another decision than we did, that's fine, but it doesn't mean we didn't sufficiently consider the options or that we would have chosen differently if only you explained it louder/harder at us.

This conversation is not being helpful to anyone at this point, so I'm locking it.

**Pull request merged**
